Core/Scripts: implement boss Apothecary Hummel ("Love is in the Air" event boss)

sirikfoll opened the pull request:
…nt boss)

**Changes proposed:**

-  Implement the fight and add the related objects

**Target branch(es):** 3.3.5/master
- [x] 3.3.5

**Issues addressed:** Closes #5114


**Tests performed:** Built


**Known issues and TODO list:** (add/remove lines as needed)

- [x] Gossip menu option don't appear, find out the correct text too (help needed)
- [x] Clean-up
- [x] Research Crazed Apothecary spawns (dropped, it was only existing in 3.3.5 and there's no sniffs of it to find out correct spawn locations)
- [x] Spells Perfume and Cologne Spill doesn't do damage (help needed)
The spells are 68798 and 68614
```
Effect 0: Id 27 (SPELL_EFFECT_PERSISTENT_AREA_AURA)
BasePoints = 250
Targets (18, 8) (TARGET_DEST_CASTER, TARGET_UNIT_DEST_AREA_ENTRY)
Aura Id 226 (SPELL_AURA_PERIODIC_DUMMY), value = 250, misc = 0 (0), miscB = 0, periodic = 1000
Radius (Id 15) 3,00
```
They have to do damage to players in range, the damage spells are 68927 and 68934
```
Effect 0: Id 2 (SPELL_EFFECT_SCHOOL_DAMAGE)
BasePoints = 2375 to 2626
Targets (25, 0) (TARGET_UNIT_TARGET_ANY, NO_TARGET)

```

sirikfoll committed:
Core/Scripts Implement boss Apothecary Hummel (Love is in the air event boss)

Temppus commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp:
```diff
@@ -0,0 +1,579 @@
+
+            void OnPeriodic(AuraEffect const* /*aurEff*/)
+            {
+                std::cout << "\n GetTarget Perfume: " << GetTarget()->GetName();
```
cleanup

Temppus commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp:
```diff
@@ -0,0 +1,579 @@
+
+            void OnPeriodic(AuraEffect const* /*aurEff*/)
+            {
+                std::cout << "\n GetTarget Cologne: " << GetTarget()->GetName();
```
cleanup

sirikfoll committed:
First clean-up

ccrs commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp:
```diff
@@ -0,0 +1,577 @@
+                {
+                    if (Group* group = players.begin()->GetSource()->GetGroup())
+                        if (group->isLFGGroup())
+                            sLFGMgr->FinishDungeon(group->GetGUID(), 288, me->GetMap());
```
Find out what causes the instance to be completed (encounters) and use it. Probably it'll be a spell.

sirikfoll replied:
afaik is a dbc problem with all those world event bosses, there's no spells involved

ccrs replied:
Umm seems like these seasonal LFGDungeons dont have any  definition in InstanceEncounters, weird

ccrs replied:
For me it would be better to add the 4 or 5 missing definitions in instance_encounters than keep doing this manual credit.
Lets see what the others think.

ccrs replied:
Talking about exactly these 4

![image](https://cloud.githubusercontent.com/assets/4329612/22753816/c1d6f242-ee3d-11e6-800d-9a89abf0edba.png)

ccrs commented on sql/updates/world/3.3.5/9999_99_99_99_world_CrownChemical.sql:
```diff
@@ -0,0 +1,234 @@
+UPDATE `creature_template` SET `unit_flags`=33280, `ScriptName`='npc_apothecary_baxter' WHERE  `entry`=36565;
+UPDATE `creature_template` SET `unit_flags`=33280, `ScriptName`='npc_apothecary_frye' WHERE  `entry`=36272;
+UPDATE `creature_template` SET `unit_flags`=33555200 WHERE  `entry` IN(36710,36530);
+UPDATE `gameobject_loot_template` SET `QuestRequired`=0 WHERE `Entry`=28677 and`Item`=49352;
```
why do you change this?

sirikfoll replied:
There's no quest that require those items, then players won't be able to loot the item

ccrs replied:
776 templates have ````QuestRequired = 1`````doesnt seem like an error

sirikfoll replied:
Sorry, didn't understood what you are trying to say

ccrs replied:
QuestRequired = true -> check where you can add that it should track the quest related to the LFGEncounter so that the items can be looted if the quest is active

sirikfoll replied:
why would I do that? This field doesn't come from WDB data, while quest_template does

Exodius commented:
@sirikfoll Dunno if this will help with spells...
http://pastebin.com/60Pjen6b

Aokromes commented:
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp:105:18: fatal error: 
      'Reset' overrides a member function but is not marked 'override'
      [-Winconsistent-missing-override]
            void Reset()
                 ^
/home/travis/build/TrinityCore/TrinityCore/src/server/game/AI/ScriptedAI/ScriptedCreature.h:367:14: note: 
      overridden virtual function is here
        void Reset() override { _Reset(); }
             ^
1 error generated.
make[2]: *** [src/server/scripts/CMakeFiles/scripts_easternkingdoms.dir/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp.o] Error 1

Kittnz commented:
Video: https://www.youtube.com/watch?v=10j3uyei90M

sirikfoll committed:
Fix Frye aggro, Hummel's loot and gossip_menu_option

sirikfoll commented:
@Exodius  It didn't work :/

Keader commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp:
```diff
@@ -112,6 +112,13 @@ class boss_apothecary_hummel : public CreatureScript
+            void EnterEvadeMode(EvadeReason /*why*/) override
+            {
+                summons.DespawnAll();
+                _EnterEvadeMode();
```
realy needed?

sirikfoll replied:
Well, since #19071 apparently will be ignored, it is. :/

sirikfoll committed:
Fix build

sirikfoll committed:
Build Fix

sirikfoll committed:
Build Fix

Viste commented:
in sql file also need spawn for alchemist don't have them in world ( tdb 62

sirikfoll commented:
@viste Did you got all the updates? It is working here, only thing missing now is the damage from the spells Cologne and Perfume Spill(which i don't know how to fix)

Viste commented:
@sirikfoll http://paste2.org/2ZD7zyUp sql http://paste2.org/fFs5VLgp cpp when use "Begin the battle." nothing change even gossip still open, but in cpp say close gossip 
`            void sGossipSelect(Player* player, uint32 menuId, uint32 gossipListId) override
            {
                if (menuId == GOSSIP_MENU_HUMMEL && gossipListId == GOSSIP_OPTION_START)
                {
                    me->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_GOSSIP);
                    CloseGossipMenuFor(player);
                    DoAction(ACTION_START_EVENT);
                }
            }`

sirikfoll commented:
@viste Did you run Cmake again before compiling? It's necessary because of the new file.
The problem could be that the script isn't loaded

Viste commented:
@sirikfoll yes ofc i not so stupid, for more clean i remove build dir and then mkdir build && cd build  and then run cmake :( 
![wowscrnshot_021217_010804](https://cloud.githubusercontent.com/assets/281977/22857885/c7bd4b78-f0bf-11e6-8d1a-c3198158da8e.jpg)
![wowscrnshot_021217_010808](https://cloud.githubusercontent.com/assets/281977/22857892/d1311e64-f0bf-11e6-8a34-8ba4c864ee9e.jpg)

sirikfoll commented:
No idea what's wrong there then :/
But it really seems that the script isn't loaded, he isn't even summoning the other 2 npcs

Viste commented:
@sirikfoll beefore i add the script and sql, no any apothecary in the inst. Fight don't start when i put the start battle, faction don't change and two other must be summoned when hummel die then when baxter die he call frye [video1](https://www.youtube.com/watch?v=FPtht0PIaL0) [video2](https://www.youtube.com/watch?v=3ynU_03GgN8)

sirikfoll committed:
Fix script loading

sirikfoll commented:
Should work now, thanks for reporting ^^

sirikfoll committed:
Add mechanic_immune_mask

Viste commented:
No problem, thank you for script!

Viste commented:
1 more report for this event if need create issue i can do this, when kill apothecary's you recive achiev and item, achiev works but item is empty http://www.wowhead.com/item=54537#contains

sirikfoll committed:
Added loot to the Heart-Shaped Box

sirikfoll commented:
Added the loot, I put the chances shown in wowhead, except for the necklaces, which I didn't find anything, if anyone have suggestions, just tell me that I change ^^

Tonghost commented on src/server/scripts/EasternKingdoms/ShadowfangKeep/boss_apothecary_hummel.cpp:
```diff
@@ -0,0 +1,587 @@
+
+            void OnPeriodic(AuraEffect const* /*aurEff*/)
+            {
+                GetTarget()->CastSpell((Unit*)nullptr, SPELL_PERFUME_SPILL_DAMAGE, true);
```
There must be SPELL_COLOGNE_SPILL_DAMAGE

Tonghost replied:
Why nullptr target? Try this
GetTarget()->CastSpell(GetTarget(), SPELL_PERFUME_SPILL_DAMAGE, true);

Keader replied:
condition select target, no?

Tonghost replied:
@Keader It would be an excellent solution

Keader replied:
Normally we use (Unit*) nullptr, when spell has conditions that select target
So if he put:
```
GetTarget()->CastSpell(GetTarget(), SPELL_PERFUME_SPILL_DAMAGE, true);
```
will not make any difference xd

Tonghost replied:
Really? Just try. This spell has no any conditions for targets.

sirikfoll replied:
Target doesn't really matter right now, the problem is that the hook isn't called at all, I'm waiting for a fix to that problem to do tests and finish the SpellScripts

Tonghost replied:
I just checked, everything worked

Viste replied:
Targets (18, 8) (TARGET_DEST_CASTER, TARGET_UNIT_DEST_AREA_ENTRY)
Targets (1, 0) (TARGET_UNIT_CASTER, NO_TARGET)

sirikfoll replied:
The damage from the spill is working? can you post a Screenshot please? Even changing targets, i can't see any damage here :/

sirikfoll committed:
Misc fixes

Improve Frye movement,
Despawn Crown Apothecary at engage,
Remove Alluring Perfume aura from Hummel at feign death
Clear EmoteState from frye
Remove stun immunity
Remove Big Love Rocket loot from Hummel, it should drop only from the Heart-shaped Box

Tonghost commented:
Now I understand what you say.
Apothecary Frye for unknown reasons has flag UNIT_FLAG_IMMUNE_TO_PC at the moment of casting spell and can not deal damage to player

Viste commented:
@sirikfoll frey still run somewhere like insane ((

sirikfoll committed:
Fix Frye Melee Attack

sirikfoll commented:
@Viste  That is correct, he should run uncontrollably, only melee attack was missing, fixed now

xinef1 commented:
Is "Spells Perfume and Cologne Spill doesn't do damage (help needed)
The spells are 68798 and 68614" Still valid? If so, try passing caster guid (should be boss in periodic dummy) as originalCasterGUID

Keader committed:
codestyle

Keader committed:
rename sql and changed for clean guids
